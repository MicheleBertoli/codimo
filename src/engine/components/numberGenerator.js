/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { Text, TextStyle } from 'pixi.js';
import { TweenLite, Linear } from 'gsap';

import { BLOCK_SIZE, HALF, ONE, ACTOR_MOVEMENT_DURATION } from 'constants/numbers';

const WIDTH = BLOCK_SIZE;
const HEIGHT = WIDTH;
const style = new TextStyle({
  fontFamily: 'Arial',
  fontSize: 38,
  fontWeight: 'bold',
  fill: ['#ffffff', '#00ff99'], // gradient
  stroke: '#4a1850',
  strokeThickness: 5,
  dropShadow: true,
  dropShadowColor: '#000000',
  dropShadowBlur: 4,
  // eslint-disable-next-line no-magic-numbers
  dropShadowAngle: Math.PI / 6,
  dropShadowDistance: 2,
});

export type NumberActor = {|
  view: Text,
  initialPosition: Array<number>,
  position: string,
  updatePosition(newPosition: string): Promise<void>,
  resetPosition(): void,
|};

export const staticNumberGenerator = (number: number) => {
  const view = new Text(number.toString(), style);

  view.anchor.x = view.anchor.y = 0.5;

  return {
    view,
  };
};

function updatePosition(newPosition: string): Promise<void> {
  // Object scope
  this.position = newPosition;
  const positionNumbers = this.position.split(',').map((string: string): number => (parseInt(string)));

  return new Promise((onComplete) => {
    TweenLite.to(this.view, ACTOR_MOVEMENT_DURATION, {
      x: positionNumbers[0] * WIDTH + (WIDTH - this.view.width) / HALF - ONE,
      y: positionNumbers[1] * HEIGHT + (HEIGHT - this.view.height) / HALF + ONE,
      ease: Linear.easeNone,
      onComplete,
    });
  });
}
function resetPosition(): void {
  this.view.x = this.initialPosition[0] * WIDTH + (WIDTH - this.view.width) / HALF - ONE;
  this.view.y = this.initialPosition[1] * HEIGHT + (HEIGHT - this.view.height) / HALF + ONE;
}
const numberGenerator = (number: number, position: string): NumberActor => {
  const view = new Text(number.toString(), style);
  const initialPosition = position.split(',').map((string: string): number => (parseInt(string)));

  view.x = initialPosition[0] * WIDTH + (WIDTH - view.width) / HALF - ONE;
  view.y = initialPosition[1] * HEIGHT + (HEIGHT - view.height) / HALF + ONE;

  return {
    view,
    initialPosition,
    position,
    updatePosition,
    resetPosition,
  };
};

export default numberGenerator;
