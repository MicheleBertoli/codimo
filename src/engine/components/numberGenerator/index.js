/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { Text, TextStyle } from 'pixi.js';
import { TweenLite, TimelineLite, Linear } from 'gsap';

import { HALF, ONE, ZERO, TWO, ANCHOR_CENTER, ACTOR_MOVEMENT_DURATION } from 'constants/numbers';
import { UnableToLeaveTheNumericLine } from 'engine/helpers/errors';
import { EASE_BE_HAPPY } from 'engine/helpers/customEases';
import { getRandomFloat } from 'engine/helpers/randomConfigurations';

const EIGHT = 8;

const styleRaw = {
  fontFamily: 'Arial',
  fontWeight: 'bold',
  fill: ['#ffffff', '#00ff99'], // gradient
  stroke: '#4a1850',
  strokeThickness: 5,
  dropShadow: true,
  dropShadowColor: '#000000',
  dropShadowBlur: 4,
  // eslint-disable-next-line no-magic-numbers
  dropShadowAngle: Math.PI / 6,
  dropShadowDistance: 2,
};

export type NumberActor = {|
  view: Text,
  position: string,
  finalPosition: string,
  beHappy(state: BeHappyState): void,
  changeActor(number: number): void,
  hasEnteredToNumericLine(): Promise<void>,
  resetPosition(): void,
  updatePosition(newPosition: string): Promise<void>,
|};
export type StaticNumberActor = {|
  view: Text,
  beHappy(state: BeHappyState): void,
|};
export type BeHappyState = 'start' | 'stop';

const beHappyConfig = (
  view: Text,
) => {
  const timeline = new TimelineLite({
    onComplete: () => {
      timeline.restart();
    },
    paused: true,
  });
  const delay = getRandomFloat(ZERO, TWO);

  timeline.to(view, ONE, {
    y: 0,
    ease: EASE_BE_HAPPY,
    delay,
  });

  return (state: BeHappyState): void => {
    if (state === 'start') {
      timeline.restart();
    } else {
      timeline.stop();
    }
  };
};

export const staticNumberGenerator = (number: number, size: number): StaticNumberActor => {
  const style = new TextStyle({
    ...styleRaw,
    fontSize: size / HALF + size / EIGHT,
  });
  const view = new Text(number.toString(), style);

  view.anchor.set(ANCHOR_CENTER);
  view.x = view.y = size / HALF;

  return {
    view,
    beHappy: beHappyConfig(view),
  };
};

const hasEnteredToNumericLineConfig = (
  view: Text,
  size: number,
  margin: number,
  /**
   * This function needs the number's scope. That's why is a named function.
   *
   * @return {Promise<void>} animation promise
   */
) => (function hasEnteredToNumericLine(): Promise<void> {
  this.position = undefined;

  view.x = size / HALF;
  view.y = size + view.x + margin + margin;

  return new Promise((onComplete) => {
    TweenLite.to(view, ACTOR_MOVEMENT_DURATION, {
      y: size / HALF,
      ease: Linear.easeNone,
      onComplete,
    });
  });
});

const updatePositionConfig = (
  view: Text,
  size: number,
  margin: number,
  /**
   * This function needs the number's scope. That's why is a named function.
   *
   * @param  {string} newPosition  to go
   * @return {Promise<void>}       animation promise
   */
) => (function updatePosition(newPosition: string): Promise<void> {
  if (!this.position) {
    throw new UnableToLeaveTheNumericLine();
  }
  this.position = newPosition;
  const positionNumbers = this.position.split(',').map((string: string): number => (parseInt(string)));

  return new Promise((onComplete) => {
    TweenLite.to(view, ACTOR_MOVEMENT_DURATION, {
      x: positionNumbers[0] * (size + margin) + size / HALF + margin,
      y: positionNumbers[1] * (size + margin) + size / HALF + margin,
      ease: Linear.easeNone,
      onComplete,
    });
  });
});
const resetPositionConfig = (
  view: Text,
  initialPosition: Array<number>,
  size: number,
  margin: number,
) => (function resetPosition() {
  this.position = `${initialPosition[0]},${initialPosition[1]}`;

  view.x = initialPosition[0] * (size + margin) + size / HALF + margin;
  view.y = initialPosition[1] * (size + margin) + size / HALF + margin;
});
const changeActorConfig = (
  view: Text,
) => (number: number) => {
  view.text = number.toString();
};

/**
 * This generator returns a new instance of NumberActor.
 *
 * TODO validate number is between valid range.
 *
 * @param  {number} number        What number will be rendered. Valid values: [-99, 99].
 * @param  {string} position      Follows the format 'x,y'.
 * @param  {string} finalPosition Follows the format 'x,y'.
 * @param  {number} size          The size of a block used as relative value
 * @param  {number} margin        Size of block's margin
 * @return {NumberActor}          Used for animate a number
 */
const numberGenerator = (
  number: number,
  position: string,
  finalPosition: string,
  size: number,
  margin: number,
): NumberActor => {
  const style = new TextStyle({
    ...styleRaw,
    fontSize: size / HALF + size / EIGHT,
  });
  const view = new Text(number.toString(), style);
  const initialPosition = position.split(',').map((string: string): number => (parseInt(string)));

  view.anchor.set(ANCHOR_CENTER);

  view.x = initialPosition[0] * (size + margin) + size / HALF + margin;
  view.y = initialPosition[1] * (size + margin) + size / HALF + margin;

  return {
    view,
    position,
    finalPosition,
    beHappy: beHappyConfig(view),
    changeActor: changeActorConfig(view),
    hasEnteredToNumericLine: hasEnteredToNumericLineConfig(view, size, margin),
    resetPosition: resetPositionConfig(view, initialPosition, size, margin),
    updatePosition: updatePositionConfig(view, size, margin),
  };
};

export default numberGenerator;
