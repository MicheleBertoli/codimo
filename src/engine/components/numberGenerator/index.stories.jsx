/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import React from 'react';

import { storiesOf } from 'test/storybook-facades';
import PixiWrapper from 'test/PixiWrapper';

import numberGenerator from './';

const POSITION = '0,0';
const SIZES = {
  one: 64,
  ten: 128,
  negativeOne: 256,
  negativeTen: 512,
};
// eslint-disable-next-line no-magic-numbers
const one = numberGenerator(1, POSITION, SIZES.one).view;
// eslint-disable-next-line no-magic-numbers
const ten = numberGenerator(10, POSITION, SIZES.ten).view;
// eslint-disable-next-line no-magic-numbers
const negativeOne = numberGenerator(-1, POSITION, SIZES.negativeOne).view;
// eslint-disable-next-line no-magic-numbers
const negativeTen = numberGenerator(-10, POSITION, SIZES.negativeTen).view;

storiesOf('engine.components.Number', module)
  .add('one digit', () => (<PixiWrapper component={one} height={SIZES.one} width={SIZES.one} />))
  .add('two digits', () => (<PixiWrapper component={ten} height={SIZES.ten} width={SIZES.ten} />))
  .add('negative one digit', () => (
    <PixiWrapper component={negativeOne} height={SIZES.negativeOne} width={SIZES.negativeOne} />
  ))
  .add('negative two digits', () => (
    <PixiWrapper component={negativeTen} height={SIZES.negativeTen} width={SIZES.negativeTen} />
  ));
