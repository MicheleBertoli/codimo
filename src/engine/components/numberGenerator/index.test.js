/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import numberGenerator from './';

describe('engine > components > numberGenerator', () => {
  it('should update its position acording to the new position', async () => {
    const rawNumber = 3;
    const initPosition = '1,1';
    const size = 64;
    const number = numberGenerator(rawNumber, initPosition, size);
    const initX = number.view.x;
    const initY = number.view.y;
    const newPosition = '2,2';

    await number.updatePosition(newPosition);

    expect(number.position).toBe(newPosition);
    expect(number.view.x).not.toBe(initX);
    expect(number.view.y).not.toBe(initY);

    number.resetPosition();

    expect(number.position).toBe(initPosition);
    expect(number.view.x).toBe(initX);
    expect(number.view.y).toBe(initY);
  });
  it('should update its position to undefined on enter to numeric line', async () => {
    const rawNumber = 3;
    const initPosition = '1,1';
    const size = 64;
    const marginBottom = 10;
    const number = numberGenerator(rawNumber, initPosition, size);
    const initX = number.view.x;
    const initY = number.view.y;

    await number.hasEnteredToNumericLine(marginBottom);

    expect(number.position).toBeUndefined();
    expect(number.view.x).not.toBe(initX);
    expect(number.view.y).not.toBe(initY);
  });
  it(
    'should throw an `UnableToLeaveTheNumericLine` if tries to call updatePosition when is in the numeric line',
    async () => {
      const rawNumber = 3;
      const initPosition = '1,1';
      const size = 64;
      const marginBottom = 10;
      const number = numberGenerator(rawNumber, initPosition, size);

      await number.hasEnteredToNumericLine(marginBottom);

      try {
        number.updatePosition('1,2');
      } catch (err) {
        expect(err.name).toBe('UnableToLeaveTheNumericLine');
        expect(err.message).toMatchSnapshot();
      }
    },
  );
  xit('should stop the animation if the promise was canceled. It should reset its position while clearing the canceled promise');
});
