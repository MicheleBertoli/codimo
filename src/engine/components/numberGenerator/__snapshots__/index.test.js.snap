// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine > components > numberGenerator should throw an \`UnableToLeaveTheNumericLine\` if tries to call updatePosition when is in the numeric line 1`] = `"Is not possible to leave the numeric line"`;
